File: testzone/agent/data/dexscreener.py
"""DexScreener API client — token price data and market info."""

import httpx
import logging
from typing import Optional

logger = logging.getLogger(__name__)


class DexScreenerClient:
    """Client for DexScreener's free API."""

    BASE_URL = "https://api.dexscreener.com/latest"

    def __init__(self):
        self.client = httpx.AsyncClient(timeout=10.0)

    async def get_token_price(self, address: str) -> Optional[dict]:
        """Get price and market data for a Solana token."""
        try:
            resp = await self.client.get(
                f"{self.BASE_URL}/dex/tokens/{address}",
            )
            resp.raise_for_status()
            data = resp.json()

            pairs = data.get("pairs", [])
            if not pairs:
                return None

            # Get the pair with highest liquidity
            best_pair = max(pairs, key=lambda p: p.get("liquidity", {}).get("usd", 0))

            return {
                "price_usd": float(best_pair.get("priceUsd", 0)),
                "price_native": float(best_pair.get("priceNative", 0)),
                "market_cap_usd": float(best_pair.get("marketCap", 0)),
                "fdv": float(best_pair.get("fdv", 0)),
                "liquidity_usd": float(best_pair.get("liquidity", {}).get("usd", 0)),
                "volume_24h": float(best_pair.get("volume", {}).get("h24", 0)),
                "volume_6h": float(best_pair.get("volume", {}).get("h6", 0)),
                "volume_1h": float(best_pair.get("volume", {}).get("h1", 0)),
                "price_change_5m": float(best_pair.get("priceChange", {}).get("m5", 0)),
                "price_change_1h": float(best_pair.get("priceChange", {}).get("h1", 0)),
                "price_change_24h": float(best_pair.get("priceChange", {}).get("h24", 0)),
                "txns_5m_buys": best_pair.get("txns", {}).get("m5", {}).get("buys", 0),
                "txns_5m_sells": best_pair.get("txns", {}).get("m5", {}).get("sells", 0),
                "pair_address": best_pair.get("pairAddress", ""),
                "dex": best_pair.get("dexId", ""),
            }

        except Exception as e:
            logger.error(f"Failed to fetch DexScreener data for {address}: {e}")
            return None

    async def search_tokens(self, query: str) -> list[dict]:
        """Search for tokens by name/symbol."""
        try:
            resp = await self.client.get(
                f"{self.BASE_URL}/dex/search",
                params={"q": query},
            )
            resp.raise_for_status()
            return resp.json().get("pairs", [])
        except Exception as e:
            logger.error(f"Failed to search DexScreener: {e}")
            return []

    async def close(self):
        await self.client.aclose()

File: projectx/src/scanner/dexscreener.ts
import { Token } from "../types/index";
import { logger } from "../utils/logger";

interface DexScreenerPair {
  chainId: string;
  dexId: string;
  url: string;
  pairAddress: string;
  baseToken: {
    address: string;
    name: string;
    symbol: string;
  };
  priceUsd: string;
  txns: {
    h24: { buys: number; sells: number };
    h6: { buys: number; sells: number };
    h1: { buys: number; sells: number };
  };
  volume: {
    h24: number;
    h6: number;
    h1: number;
  };
  priceChange: {
    h24: number;
    h6: number;
    h1: number;
  };
  liquidity: {
    usd: number;
  };
  fdv: number;
  marketCap: number;
  pairCreatedAt: number;
  info?: {
    imageUrl?: string;
    websites?: { url: string }[];
    socials?: { type: string; url: string }[];
  };
}

export class DexScreenerClient {
  private lastFetchTime = 0;
  private fetchDelay = 1000;
  private cache: Map<string, { data: Partial<Token>; timestamp: number }> = new Map();
  private cacheTTL = 30000; // 30 seconds

  async getTokenData(mint: string): Promise<Partial<Token> | null> {
    // Check cache first
    const cached = this.cache.get(mint);
    if (cached && Date.now() - cached.timestamp < this.cacheTTL) {
      return cached.data;
    }

    try {
      const now = Date.now();
      const timeSinceLastFetch = now - this.lastFetchTime;
      if (timeSinceLastFetch < this.fetchDelay) {
        await new Promise(resolve => setTimeout(resolve, this.fetchDelay - timeSinceLastFetch));
      }

      const url = `https://api.dexscreener.com/tokens/v1/solana/${mint}`;
      const response = await fetch(url);
      this.lastFetchTime = Date.now();

      if (!response.ok) {
        return null;
      }

      const pairs = await response.json() as DexScreenerPair[];
      if (!pairs || pairs.length === 0) {
        return null;
      }

      // Find the pair with highest liquidity
      const bestPair = pairs.sort((a, b) => (b.liquidity?.usd || 0) - (a.liquidity?.usd || 0))[0];

      const data: Partial<Token> = {
        volume24h: bestPair.volume?.h24 || 0,
        liquidity: bestPair.liquidity?.usd || 0,
        buys24h: bestPair.txns?.h24?.buys || 0,
        sells24h: bestPair.txns?.h24?.sells || 0,
        txns24h: (bestPair.txns?.h24?.buys || 0) + (bestPair.txns?.h24?.sells || 0),
        marketCap: bestPair.marketCap || bestPair.fdv || 0,
        price: parseFloat(bestPair.priceUsd) || 0,
        twitter: bestPair.info?.socials?.find(s => s.type === "twitter")?.url,
        telegram: bestPair.info?.socials?.find(s => s.type === "telegram")?.url,
        website: bestPair.info?.websites?.[0]?.url,
      };

      // Cache the result
      this.cache.set(mint, { data, timestamp: Date.now() });

      return data;
    } catch (error) {
      logger.error(`DexScreener fetch failed for ${mint}:`, error);
      return null;
    }
  }

  async enrichTokens(tokens: Token[]): Promise<Token[]> {
    const enriched: Token[] = [];

    for (const token of tokens) {
      const extra = await this.getTokenData(token.mint);
      if (extra) {
        enriched.push({ ...token, ...extra });
      } else {
        enriched.push(token);
      }
    }

    return enriched;
  }
}
